client/src/research/ResearcherDataPage.js

```diff
@@ -193,6 +191,41 @@ class Analysis extends Component {
       .catch(this.onError.bind(this));
   }
 
+  getAudioID(audioUrl) {
+    if (audioUrl) {
+      const slashIndex = audioUrl.lastIndexOf('/');
+      const filename = audioUrl.slice(slashIndex + 1);
+      return filename;
+    }
+    return "";
+  }
+  getAudio(audioID,elementID) {
+    const token = this.state.token;
+
+    fetch('/server/research/wav/'+audioID, {
+      headers: {
+        'Accept': 'application/json',
+        'Content-Type': 'application/json',
+        'x-teachermoments-token': token,
+      },
+      method: 'GET'
+    })
+      .then(results => {
+        var response = new Response(results.body, {headers: {"Content-Type": "audio/wav"}});
+        response.blob().then(function(myBlob) {
+          var bloburl = URL.createObjectURL(myBlob);
+          var elementTarget = document.getElementById(audioID.slice(0, -4));
+          if (elementTarget) {
+            document.getElementById(audioID.slice(0, -4)).src = bloburl;
```

Comment:
As a rule of thumb, you shouldn't ever be calling `document.getElementById` in React components.  If you really really need to work directly with a DOM element in React, you can use `refs`. But usually it's better to call `setState` with what new data or information you got and then let `render` read that data and convert it to HTML.  Here, if you keep this approach to getting audio blobs, you would take the new blob reference, store it in the component's state, and then express what that should look like in HTML in `render` methods.

Separately, I think it's also worth exploring if you can directly render an `<audio />` tag with the URL set to the server URL for a particular `audioID`, or understanding where that's stuck.  That's what the private repo did previously, and that would let you remove all this code.  That may not work if it doesn't let you send the token in the header along with the request.  If that's the case another way to improve this would be to exchange the token with the server for a cookie on login, since I believe the browser will send cookies with `<audio />` requests.

Regardless this is all another reason to factor all this complexity out into its own `<AudioWithTranscription />` component or something like that, and remove the DOM manipulation code (related to https://github.com/mit-teaching-systems-lab/threeflows/pull/348/files#r180895553).

Reply:
Oh yes. This was my last ditch effort to get this done before the Friday playtest. I was planning on the audioplayers into their own component. I made issue https://github.com/mit-teaching-systems-lab/threeflows/issues/362 to make sure I do this